tfx/components/example_gen/utils.py

```diff
@@ -209,51 +213,96 @@ def _glob_to_regex(glob_pattern: Text) -> Text:
+      raise ValueError('Cannot not find %s number from %s based on %s' %
+                       (property_name, file_path, split_regex_pattern))
+    if latest_property is None or prop >= int(latest_property):
       # Uses str instead of int because of zero padding digits.
-      latest_span = result.group(1)
+      latest_property = result.group(1)
 
-  if latest_span is None:
+  if latest_property is None:
     raise ValueError('Cannot not find matching for split %s based on %s' %
                      (split.name, split.pattern))
-  return latest_span
+  return latest_property
+
+
+def _retrieve_latest_span_version(uri: Text,
+                                  split: example_gen_pb2.Input.Split
+                                  ) -> Tuple[Text, Text]:
+  """Retrieves the most recent span and version for a given split pattern"""
+  split_pattern = os.path.join(uri, split.pattern)
+
+  latest_span = '0'
+  latest_version = '0'
+
+  if SPAN_SPEC in split.pattern:
+    if split.pattern.count(SPAN_SPEC) != 1:
+      raise ValueError('Only one {SPAN} is allowed in %s' % split_pattern)
+
+    split_glob_pattern = split_pattern.replace(SPAN_SPEC, '*')
+    split_glob_pattern = split_glob_pattern.replace(VERSION_SPEC, '*')
+    logging.info('Span glob pattern for split %s: %s', split.name,
+                 split_glob_pattern)
+    split_regex_pattern = _glob_to_regex(split_pattern).replace(SPAN_SPEC,
+                                                                '(.*)')
+    split_regex_pattern = split_regex_pattern.replace(VERSION_SPEC, '.*')
+    logging.info('Span regex pattern for split %s: %s', split.name,
+                 split_regex_pattern)
 
+    latest_span = _property_search(split, split_glob_pattern,
+                                   split_regex_pattern, SPAN_PROPERTY_NAME)
 
-def calculate_splits_fingerprint_and_span(
+  if VERSION_SPEC in split.pattern:
+    if split.pattern.count(VERSION_SPEC) != 1:
+      raise ValueError('Only one {VERSION} is allowed in %s' % split_pattern)
+
+    split_glob_pattern = split_pattern.replace(SPAN_SPEC, latest_span)
+    split_glob_pattern = split_glob_pattern.replace(VERSION_SPEC, '*')
+    logging.info('Version glob pattern for split %s: %s', split.name,
+                 split_glob_pattern)
+    split_regex_pattern = _glob_to_regex(split_pattern).replace(SPAN_SPEC,
+                                                                latest_span)
+    split_regex_pattern = split_regex_pattern.replace(VERSION_SPEC, '(.*)')
+    logging.info('Version regex pattern for split %s: %s', split.name,
+                 split_regex_pattern)
+
+    latest_version = _property_search(split, split_glob_pattern,
+                                      split_regex_pattern,
+                                      VERSION_PROPERTY_NAME)
+
+  split.pattern = split.pattern.replace(SPAN_SPEC, latest_span)
+  split.pattern = split.pattern.replace(VERSION_SPEC, latest_version)
+
+  return latest_span, latest_version
```

Comment:
can we use up to two capture groups? 
if span is given, but not version, we use one for span, return span number, and nil for version
if span/version are given, we use two,  return a pair of numbers, error out if number is invalid
if span is not given, version is given, we error out. 

by doing so, we avoid glob twice, and we can combine the _property_search in a cleaner way to update the tuple together.

Reply:
the main rationale for my decision to do glob twice is I was thinking about edge cases like if we had a pattern `{VERSION}/{SPAN}/...`. is assuming an ordering of span then version in the split config something we can do?

Reply:
after working at this for a bit, i realize that we can use named groups to specifically parse our span and version, regardless of the order. fixed it so that only one glob pass is needed.